src/light_data/light_store.rs

```diff
@@ -0,0 +1,404 @@
+use proptest_derive::Arbitrary;
+
+use anyhow::anyhow;
+use std::collections::BTreeMap;
+
+use crate::light_data::Encodable;
+use crate::light_data::LightData;
+use crate::scalar_store::ScalarExpression;
+use crate::scalar_store::ScalarStore;
+use crate::store::ScalarPointer;
+use crate::store::ScalarPtr;
+use crate::sym::Sym;
+use crate::tag::ExprTag;
+
+use crate::field::LurkField;
+
+#[derive(Debug, PartialEq)]
+#[cfg_attr(not(target_arch = "wasm32"), derive(Arbitrary))]
+#[cfg_attr(not(target_arch = "wasm32"), proptest(no_bound))]
+/// LightStore contains a fragment of the ScalarStore, but using the `LightExpr` type
+pub struct LightStore<F: LurkField> {
+    /// An analogous to the ScalarStore's scalar_map, but with `LightExpr` instead of `ScalarExpression`
+    pub scalar_map: BTreeMap<ScalarPtr<F>, Option<LightExpr<F>>>,
+}
+
+impl<F: LurkField> Encodable for LightStore<F> {
+    fn ser(&self) -> LightData {
+        // TODO: this clone is loathsome
+        self.scalar_map
+            .clone()
+            .into_iter()
+            .collect::<Vec<(ScalarPtr<F>, Option<LightExpr<F>>)>>()
+            .ser()
+    }
+    fn de(ld: &LightData) -> anyhow::Result<Self> {
+        let pairs = Vec::<(ScalarPtr<F>, Option<LightExpr<F>>)>::de(ld)?;
+        Ok(LightStore {
+            scalar_map: pairs.into_iter().collect(),
+        })
+    }
+}
+
+impl<F: LurkField> LightStore<F> {
+    fn insert_scalar_string(
+        &self,
+        ptr0: ScalarPtr<F>,
+        store: &mut ScalarStore<F>,
+    ) -> anyhow::Result<String> {
+        let mut s = String::new();
+        let mut tail_ptrs = vec![];
+        let mut ptr = ptr0;
+        let strnil_ptr = ScalarPtr::from_parts(ExprTag::Str.as_field(), F::zero());
+
+        // TODO: this needs to bail on encountering an opaque pointer
+        while let Some(LightExpr::StrCons(c, cs)) = self.get(&ptr).flatten() {
+            let chr = c
+                .value()
+                .to_char()
+                .ok_or_else(|| anyhow!("Non-char head in LightExpr::StrCons"))?;
+            store.insert_scalar_expression(c, Some(ScalarExpression::Char(chr)));
+            s.push(chr);
+            if cs != strnil_ptr {
+                tail_ptrs.push(cs);
+            }
+            ptr = cs
+        }
+        // Useful when called from insert_scalar_symbol
+        if s.is_empty() {
+            return Err(anyhow!("encountered no StrCons in LightStore::insert_scalar_string, is this a tring pointer?"));
+        }
+
+        // If we've already inserted this string, no need to go through suffixes again
+        if let Some(ScalarExpression::Str(old_value)) = store
+            .insert_scalar_expression(ptr0, Some(ScalarExpression::Str(s.clone())))
+            .flatten()
+        {
+            if old_value == s {
+                return Ok(s);
+            }
+        }
+        store.insert_scalar_expression(strnil_ptr, Some(ScalarExpression::Str(String::from(""))));
```

Comment:
If I understand this (and I might not), this is garbage-in/garbage-out, and these methods could be used to create a bad `ScalarStore` (i.e. with incorrect hashes), if the `LightStore` is also wrong, right? Obviously we don't want to pay a hashing cost for simple conversion, but it might be worthwhile at some point to have a validating version of this conversion.

Reply:
The current methods are private and can only be called on a `LightStore`. They cannot be called outside of this context.
With that said, `insert_scalar_expression` can. I'll see if I can clamp down on the visibility of this method.

To your point about the future, the `LightExpr` type will replace `ScalarExpression`, making the conversion moot.